packaging/generate-docs.py

```diff
@@ -0,0 +1,82 @@
+        explanation = "all traits with their properties and their default values plus developer commentary"
+    elif collectionName == "WeaponTypes":
+        explanation = "a template for weapon definitions as well as its contained types (warheads and projectiles) with default values and developer commentary"
+    elif collectionName == "SequenceTypes":
+        explanation = "all sequence types with their properties and their default values plus developer commentary"
+
+    outputFile = open(args.output, "w")
+    outputFile.write(f"This documentation is aimed at modders and has been automatically generated for version `{version}` of OpenRA. " +
+				"Please do not edit it directly, but instead add new `[Desc(\"String\")]` tags to the source code.\n\n")
+
+    outputFile.write(f"Listed below are {explanation}:\n\n")
+
+    for namespace in typesByNamespace:
+        outputFile.write(f'## {namespace}\n')
+        outputFile.write('\n')
+
+        for currentType in typesByNamespace[namespace]:
+            outputFile.write(f'### {currentType["Name"]}\n')
+            outputFile.write(f'#### {currentType["Description"]}\n')
+
+            if "InheritedTypes" in currentType and currentType["InheritedTypes"]:
+                outputFile.write("Inherits from: " + ", ".join([f'[`{x}`](#{x.lower()})' for x in currentType["InheritedTypes"]]) + '.\n\n')
+
+            if "RequiresTraits" in currentType and currentType["RequiresTraits"]:
+                outputFile.write("Requires trait(s): " + ", ".join([f'[`{x}`](#{x.lower()})' for x in currentType["RequiresTraits"]]) + '.\n')
+
+            if len(currentType["Properties"]) > 0:
+                hasAttributes = "OtherAttributes" in currentType["Properties"][0]
+                outputFile.write('\n')
+                outputFile.write(f'| Property | Default Value | Type |{" Attributes |" if hasAttributes else ""} Description |\n')
+                outputFile.write(f'| -------- | ------------- | ---- |{" ---------- |" if hasAttributes else ""} ----------- |\n')
+
+                for prop in currentType["Properties"]:
+                    if "OtherAttributes" in prop:
+                        attributesList = [f'{x["Name"]}{"(" + str(x["Value"][0]) + ")" if x["Value"] else ""}' for x in prop["OtherAttributes"] if x["Name"]]
+                        if not attributesList:
+                            attributes = ''
+                        else:
+                            attributes = ", ".join(attributesList)
+
+                        outputFile.write(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["Type"]} | {attributes} | {prop["Description"]} |\n')
+                    else:
+                        outputFile.write(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["Type"]} | {prop["Description"]} |\n')
+
+            outputFile.write('\n#\n\n')
+
+    outputFile.close()
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
```

Comment:
IMO it would be cleaner if we called this file `format-docs.py` and had it read directly from stdin and write directly to stdout instead of using temporary files. The workflow would then run

```bash
./utility.sh all --docs "${GIT_TAG}" | python3 ./packaging/format-docs.py > docs/api/playtest/traits.md
```

Reply:
Should be done now.